### PR TITLE
Example test for Frozen Bitmap fault when using Frozen Bitmaps as regular/mutable bitmaps with a read-only buffer

### PR DESCRIPTION
**Pull request by jacksonrnewhouse**
This is a test that should not pass but does, as it asserts a panic exactly when FrozenBitmaps and MMap is used. @a392n328 helped me diagnose these issues.

### CODE REVIEW

File: serialization_frozen_test.go [hunk 1]
```diff
@@ -126,3 +131,90 @@ func TestFrozenFormat(t *testing.T) {
...
+		require.NoError(t, err)
+		return receiver
+	}
+	_, err = receiver.FromBuffer(nonFrozenBytes)
+	require.NoError(t, err)
+	return receiver
+}
+
+func getBytesAsMMap(t *testing.T, startingBytes []byte) []byte {
+	os.WriteFile("data", startingBytes, 0777)
```
Comment (lemire) [inline serialization_frozen_test.go:210]: Surely, there is some needed error handling here?

Note that os.WriteFile is go 1.16. I think you want ioutil.WriteFile for portability.

File: serialization_frozen_test.go [hunk 2]
```diff
@@ -126,3 +131,90 @@ func TestFrozenFormat(t *testing.T) {
...
+func testFrozenCase(t *testing.T, testCase frozenTestCase) {
+
+	startingBitmap := NewBitmap()
+	startingBitmap.highlowcontainer.appendContainer(0, &runContainer16{iv: []interval16{{0, 0},
+		{2, 5}}}, false)
+
+	primary := getBitmap(t, testCase, startingBitmap)
+	assert.True(t, startingBitmap.Equals(primary))
+	clone := primary.Clone()
+	primary.Xor(clone)
```
Comment (lemire) [inline serialization_frozen_test.go:171]: You are modifying, in-place, the frozen view.

File: serialization_frozen_test.go [hunk 3]
```diff
@@ -126,3 +131,90 @@ func TestFrozenFormat(t *testing.T) {
...
+	res := primary.GetCardinality()
+	assert.Equal(t, res, uint64(0))
+	if testCase.shouldPanic {
+		assert.Panics(t, func() {
+			debug.SetPanicOnFault(true)
+			primary.Xor(clone)
```
Comment (lemire) [inline serialization_frozen_test.go:177]: You are, again, modifying the frozen view.

File: serialization_frozen_test.go [hunk 4]
```diff
@@ -126,3 +131,90 @@ func TestFrozenFormat(t *testing.T) {
...
+}
+
+func getBytesAsMMap(t *testing.T, startingBytes []byte) []byte {
+	os.WriteFile("data", startingBytes, 0777)
+	file, err := os.OpenFile("data", os.O_RDONLY, 0)
+	require.NoError(t, err)
+	fileinfo, err := file.Stat()
+	require.NoError(t, err)
+
+	bytes, err := unix.Mmap(int(file.Fd()), 0, int(fileinfo.Size()),
```
Comment (lemire) [inline serialization_frozen_test.go:216]: There is a ressource leak around it, surely.

**Comment by lemire** [comment]
@Oppen Could you take a look?

Users are reminded that the frozen support in roaring is experimental. It has minimal testing, and there are no tests with mutation over frozen bitmaps. The frozen format was designed for immutable bitmaps (that you just access without modifying). Using frozen bitmaps as regular bitmaps is an extension.

File: serialization_frozen_test.go (hunk 3 above)
Comment (lemire) [reply serialization_frozen_test.go:177]: I would do 

```
newprimary := Xor(primary, clone)
```

I am not sure why you want to mutate the frozen view.

**Comment by lemire** [comment]
This PR cannot be merged. I converted it to a draft.

File: serialization_frozen_test.go [hunk 5]
```diff
@@ -126,3 +131,90 @@ func TestFrozenFormat(t *testing.T) {
...
+		unix.PROT_READ, unix.MAP_PRIVATE)
```
Comment (lemire) [inline serialization_frozen_test.go:217]: `PROT_READ` means that the data can be *read*, it cannot be written to.

**Comment by lemire** [comment]
@jacksonrnewhouse How sure are you that the byte array you get from the memory map is never written to when you mutate the bitmap? Your `byte[]` has to be read-only.

File: serialization_frozen_test.go [hunk 6]
```diff
@@ -126,3 +131,90 @@ func TestFrozenFormat(t *testing.T) {
...
+		return receiver
+	}
+	_, err = receiver.FromBuffer(nonFrozenBytes)
+	require.NoError(t, err)
+	return receiver
+}
+
+func getBytesAsMMap(t *testing.T, startingBytes []byte) []byte {
+	os.WriteFile("data", startingBytes, 0777)
+	file, err := os.OpenFile("data", os.O_RDONLY, 0)
```
Comment (lemire) [inline serialization_frozen_test.go:211]: The file is read-only.

File: serialization_frozen_test.go (hunk 4 above)
Comment (lemire) [inline serialization_frozen_test.go:216]: bytes has to be immutable, it cannot be written to.

**Comment by Oppen** [comment]
As @lemire pointed out, frozen bitmaps are not meant to be modified. The name attempts to make that clear. It could maybe handle that better or behave as CoW, but that's arguably a bad idea, as you'd potentially be duplicating memory use and inducing a leak. The idea of frozen bitmaps is to have an optimized version of read only bitmaps (in particular one that is compatible with CRoaring).

I'll look into it with more detail at night.

**Comment by lemire** [comment]
I am going to close this.

**Pull request closed without merging**